sdk/python/kfp/dsl/legacy_data_passing_adaptor.py

```diff
@@ -0,0 +1,144 @@
+passing. Passing parameters to artifacts or vice versa could result unexpected
+runtime errors. It is highly recommended to not rely on the adaptors. We may
+remove the adaptors in the later releases.
+"""
+from typing import Type, Union
+
+from kfp import dsl
+from kfp.components import _naming
+from kfp.dsl import dsl_utils
+from kfp.dsl import io_types
+from kfp.dsl import type_utils
+from kfp.pipeline_spec import pipeline_spec_pb2 as pb
+
+_PARAMETER_TYPES = Union[str, int, float, bool, dict, list]
+
+# Global index to generate unique and stable adaptor name.
+adaptor_op_index = 0
+
+
+class BaseAdaptor:
+  """Base class of legacy data passing adaptor.
+
+  Attributes:
+    INPUT_KEY: The name of the component input.
+    OUTPUT_KEY: The name of the component output.
+    task_spec: The IR PipelineTaskSpec of the component.
+    component_spec: The IR ComponentSpec of the component.
+    container_spec: The IR PipelineContainerSpec of the component.
+  """
+  INPUT_KEY = 'input'
+  OUTPUT_KEY = 'output'
+  _IMAGE = 'alpine:3.14'
+
+  def __init__(self, base_name: str):
+    """Inits partially the task_spec.
+
+    The task input spec will be assigned during the components connecting phase
+    (in `kfp.dsl._component_bridage`).
+
+    Args:
+      base_name: The base name of the adaptor op. Is will be suffixed with a
+        unique index to make the component name unique within a pipeline.
+    """
+    global adaptor_op_index
+    self._name = f'{base_name}-{adaptor_op_index}'
+    adaptor_op_index += 1
+
+    self.task_spec = pb.PipelineTaskSpec()
+    self.task_spec.task_info.name = dsl_utils.sanitize_task_name(self._name)
+    self.task_spec.component_ref.name = dsl_utils.sanitize_component_name(
+        self._name)
+
+
+class ParameterToArtifactAdaptor(BaseAdaptor):
+  """The adaptor to convert a parameter into an artifact."""
+
+  def __init__(
+      self,
+      input_type: _PARAMETER_TYPES,
+      output_type: Union[str, Type[io_types.Artifact]],
+  ):
+    """Inits the component_spec, the container_spec, and part of the task_spec.
+
+    Args:
+      input_type: The type of the component input. Expects a parameter type.
+      output_type: The type of the component output. Expects an artifact type.
+    """
+    super().__init__('legacy-p2a-adaptor')
+
+    self.component_spec = pb.ComponentSpec()
+    self.component_spec.input_definitions.parameters[
+        self.INPUT_KEY].type = type_utils.get_parameter_type(input_type)
+    self.component_spec.output_definitions.artifacts[
+        self.OUTPUT_KEY].artifact_type.CopyFrom(
+            type_utils.get_artifact_type_schema(output_type))
+    self.component_spec.executor_label = dsl_utils.sanitize_executor_label(
+        self._name)
+
+    self.container_spec = pb.PipelineDeploymentConfig.PipelineContainerSpec(
+        image=self._IMAGE,
+        command=[
+            'sh',
+            '-c',
+            '-ex',
+            'mkdir -p "$(dirname "$1")"; echo "$0" > "$1"',
```

Comment:
Unfortunately, echo mangles the data (e.g. `\n` sequences).
Use `printf "%s" "$0"` instead of `echo "$0"`.